### PR TITLE
Improve map<json> type checking

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Improve compile time type checking to allow assign json compatible record to be assigned to map of json.

```ballerina
public function testCloseRecordToMapJsonAssigment() returns [map<json>, map<json>] {
    AnotherPerson ap = {};
    Person p = {};
    map<json> pm = p;
    map<json> m = ap; // was a compile time error: incompatible types: expected 'map<json>', found 'AnotherPerson'

    return [m, pm]; // was runtime error: {ballerina/lang.array}InherentTypeViolation message=incompatible types: expected 'map<json>', found 'AnotherPerson'
}

// all member types including rest type is json compatible
type Person record {|
    string name = "";
    int age = 10;
    string...;
|};

// all member types are json compatible
type AnotherPerson record {|
    string name = "";
    int age = 10 + 20;
|};
```

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/18705 and https://github.com/ballerina-platform/ballerina-lang/issues/19260

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -622,9 +621,17 @@ private boolean isAssignable(BType source, BType target, List<TypePair> unresolv
...
     }
 
-    private boolean isAssignableRecordType(BRecordType recordType, BMapType mapType) {
-        return  recordType.fields.stream().allMatch(field -> isAssignable(field.type, mapType.constraint)) &&
-                isAssignable(recordType.restFieldType, mapType.constraint);
+    private boolean isAssignableRecordType(BRecordType recordType, BMapType targetMapType) {
+        if (recordType.sealed) {
+            return recordFieldsAssignableToMap(recordType, targetMapType);
+        } else {
+            return  isAssignable(recordType.restFieldType, targetMapType.constraint)
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]:
```suggestion
            return isAssignable(recordType.restFieldType, targetMapType.constraint)
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/jsontype/json-value-negative.bal [hunk 2]
```diff
@@ -13,3 +13,39 @@ function testJsonInitWithUnsupportedtypes() returns (json) {
...
+type PersonWithTypedesc record {
+    string name = "";
+    typedesc<any> t;
+};
+
+type Obj object {
+    int i;
+
+    function __init() {
+		self.i = 0;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/jsontype/json-value-negative.bal]: formatting issue.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java [hunk 3]
```diff
@@ -493,6 +497,15 @@ private static boolean checkIsMapType(BType sourceType, BMapType targetType, Lis
         }
     }
 
+    private static boolean isJsonCompatibleMap(List<BType> types, BType jsonType) {
+        for (BType type : types) {
+            if (!checkIsJSONType(type, (BJSONType) jsonType, new ArrayList<>())) {
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java]: +1 for removing the second param.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java [hunk 4]
```diff
@@ -481,10 +482,13 @@ private static boolean checkIsMapType(BType sourceType, BMapType targetType, Lis
                 if (!recType.sealed) {
                     types.add(recType.restFieldType);
                 }
+                if (targetConstrainedType.getTag() == TypeTags.JSON_TAG) {
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java]: My original idea was to exit the check early, but I like your idea, updated.

**Pull request merged**